docs/communicate-with-javascript.md

```diff
@@ -1669,3 +1670,485 @@ nullable]` above. Currently 4 directives are supported: `null_to_opt`,
+also takes a last argument of type `unit`. This non-labeled argument is
+necessary to allow the compiler to determine when the optional argument has been
+omitted during function application. Further details about optional labeled
+arguments can be found in [the corresponding section of the OCaml
+manual](https://v2.ocaml.org/manual/lablexamples.html#s:optional-arguments).
+
+The functions `nameGet` and `ageGet` are accessors for each record field:
+
+```ocaml
+let twenty = ageGet alice
+
+let bob = nameGet bob
+```
+
+This generates:
+
+```javascript
+var twenty = alice.age;
+
+var bob = bob.name;
+```
+
+The functions are named by appending `Get` to the field names of the record to
+prevent potential clashes with other values within the module. If shorter names
+are preferred for the getter functions, there is an alternate `{ abstract =
+light }` payload that can be passed to `bs.deriving`:
+
+```ocaml
+type person = {
+  name : string;
+  age : int;
+}
+[@@bs.deriving { abstract = light }]
+
+let alice = person ~name:"Alice" ~age:20
+let aliceName = name alice
+```
+
+Which generates:
+
+```javascript
+var alice = {
+  name: "Alice",
+  age: 20
+};
+
+var aliceName = alice.name;
+```
+
+In this example, the getter functions share the same names as the object fields.
+Another distinction from the previous example is that the `person` constructor
+function no longer requires the final `unit` argument since we have excluded the
+optional field in this case.
+
+##### Compatibility with OCaml features
+
+The `@bs.deriving abstract` attribute and its lightweight variant can be used
+with [mutable
+fields](https://v2.ocaml.org/manual/coreexamples.html#s:imperative-features) and
+[private types](https://v2.ocaml.org/manual/privatetypes.html), which are
+features inherited by Melange from OCaml.
+
+When the record type has mutable fields, Melange will generate setter functions
+for them. For example:
+
+```ocaml
+type person = {
+  name : string;
+  mutable age : int;
+}
+[@@bs.deriving abstract]
+
+let alice = person ~name:"Alice" ~age:20 
+
+let () = ageSet alice 21
+```
+
+This will generate:
+
+```javascript
+var alice = {
+  name: "Alice",
+  age: 20
+};
+
+alice.age = 21;
+```
+
+If the `mutable` keyword is omitted from the interface file, Melange will not
+include the setter function in the module signature, preventing other modules
+from mutating any values from the type.
+
+Private types can be used to prevent Melange from creating the constructor
+function. For example, if we define `person` type as private:
+
+```ocaml
+type person = private {
+  name : string;
+  age : int;
+}
+[@@bs.deriving abstract]
+```
+
+The accessors `nameGet` and `ageGet` will still be generated, but not the
+constructor `person`. This is useful when binding to JavaScript objects while
+preventing any Melange code from creating values of such type.
+
+### Create JavaScript objects using external functions
```

Comment:
I am not sure where this section should be. It is not `bs.deriving` but it is kind of "code generation". One alternative would be to move it to "Bind to JS objects" section... 🤔

Reply:
I'd probably move it to the "Bind to JS objects" section, indeed.

Reply:
https://github.com/melange-re/documentation-site/pull/28/commits/987495dc2496d926b22108832a2b38a342511b1f.